### PR TITLE
#1612 - Add or Replace File via API

### PR DESCRIPTION
**Pull request by raprasad**
# RFI Checklist

_**Before** submitting the pull request, fill out sections (1.) Related Issues and (2.) Pull Request Checklist._
### 1. Related Issues
# 1612 Allow file uploads for non-zip files through the API
---
### 2. Pull Request Checklist
- [x]  Functionality completed as described in FRD
- [x]  Dependencies, risks, assumptions in FRD addressed
- [x]  Unit tests completed
- [ ] REST Assured tests added or modified: DatasetsIT, FilesIT
- [x]  Deployment requirements identified (e.g., SQL scripts, indexing)
- [x]  Documentation completed
- [x]  All code checkins completed
---
### 3. Review Checklist

_**After** the pull request has been submitted, fill out this section._
- [ ]  Code review completed or waived
- [ ]  Testing requirements completed
- [ ]  Usability testing completed or waived
- [ ]  Support testing completed or waived
- [ ]  Merged with develop branch and resolved conflicts

Connects to #1612

### CODE REVIEW

**Comment by raprasad** [comment]
For ticket #1612

**Comment by djbrooke** [comment]
Connects to #1612

**Comment by raprasad** [comment]
In conversation @pdurbin noted that this feature would be more complete with:
  - **API delete file** - functionality
  
This makes complete sense from a user point of view.  Since completing this ticket is a "side effect" of working on file replace (#2290), it would be good to get **API Add** in while we have the opportunity--even if **API delete file** is not included.

In the ideal situation, **API delete file** functionality would be included in the same release (4.6) as this pull request.  e.g. if time allows can work with @pdurbin on that.

**Comment by pdurbin** [comment]
@raprasad yes! Aren't you also working on being able to set descriptions of files? If so, we should document how to do this in the API Guide and put #723 through QA explaining that we have no plans to support setting descriptions via SWORD but there is a new way to do it in the native API.

**Comment by pdurbin** [comment]
@raprasad @djbrooke for supporting the deletion of files via native API I just opened #3419.

**Comment by raprasad** [comment]
@pdurbin : adding an API for to add descriptions/tags to existing files is low-hanging fruit but am avoiding scope creep for now.  (literally thought through the code line by line but wrote nothing and have now blanked it out:)

**Comment by raprasad** [comment]
fyi: found a bug with ADD.

REPLACE is retricted to a single file--even if a single '.zip' is "unzipped into multiple files.
ADD should not have this restriction.

Fix:
 - [ ] Allow add of single file (e.g. zip) to result in multiple files in dataset
 - [ ] Return appropriate JSON listing all files returned

File: src/main/java/edu/harvard/iq/dataverse/api/ApiConfiguration.java [hunk 1]
```diff
@@ -1,8 +1,18 @@
 package edu.harvard.iq.dataverse.api;
 
 import javax.ws.rs.ApplicationPath;
-import javax.ws.rs.core.Application;
+import org.glassfish.jersey.media.multipart.MultiPartFeature;
+import org.glassfish.jersey.server.ResourceConfig;
```
Comment (pdurbin) [inline src/main/java/edu/harvard/iq/dataverse/api/ApiConfiguration.java]: @scolapasta are we consciously switching from a JAVA EE standard (JAX-RS) to an specific implementation (Jersey)?

**Comment by raprasad** [comment]
# Post code review checklist

- [ ] For file add/replace remove ability to add DataFile tags
  - Leave "logic" in place to use for in #3422 - changing label, description and tags of an existing file
    - For existing files, we can validate whether the file is tabular and can have a DataFile tag
- [ ] Use native JSON.  e.g. Add new attributes to the native JSON, exclude dataset and dataverse information
  - [ ] rework ```okResponseGsonObject``` to use the native JSON object builder.  
    - This is still a new method which take both a ```JsonObjectBuilder``` object _and_ a String ```message```
- [ ] For add file API endpoint, we now have the dataset object--pass directly to AddReplaceHelper instead of the id
  - [ ] Redo business logic
  - [ ] Redo tests
- [ ] Tech debt: Update existing ```findDatasetOrDie``` method to use error messages in bundles
  - [ ] Rename existing bundle errors
  - [ ] Update add/replace business logic to use renamed bundle errors  
  - [ ] Update API tests to use renamed bundle errors
- [ ] Move ```removeDuplicatesNullsEmptyStrings``` to a utility class
  - _Which one?_
- [ ] For ```OptionalFileParams```, rename **tags** to **categories** in code/methods
  - [ ] Rename business logic methods
  - [ ] Update tests
  - There's an inconsistency where the UI/user facing logic has the word **tags** but internal code uses **categories**
     - Group decision to keep **categories** in the code
- [ ] ```AddReplaceHelper``` logic  
   - Consolidate calls to ```UpdateDatasetCommand``` in ```step_070_run_update_dataset_command``` and ```step_080_run_update_dataset_command_for_replace```

---

## Include in this ticket?

- [ ] Rename the ```ingestService.addFiles(..)``` method to ```ingestService.saveFiles(..)```
  - retest: DatasetPage UI, EditDataFiles UI, and MediaResourceManagerImpl.java

## Important Prov. note related #3238

- The add/replace API endpoints currently have a "pathway" to add additional JSON variables.  e.g. If you're adding string, numeric, boolean or other JSON based data
- However, the API endpoints will need the ability to add a second file--a prov file.

**Comment by raprasad** [comment]
This is frozen. Dev has switched to ```2290-file-replace```

**Pull request closed without merging**